aws-samples/aws-serverless-airline-booking
# EP5 and EP6 - Building and Testing Loyalty service

*Issue #, if available:*

*Description of changes:*

## Ep5 and Ep6

**Recording**

* Ep4 - **Building** the Loyalty service: https://www.youtube.com/watch?v=15YQS51Sv44&list=PLhr1KZpdzukcYWC1xD-vidMZf2uilGko
* Ep5 - **Testing** the Loyalty service: https://www.youtube.com/watch?v=zSy8KPCstJU&list=PLhr1KZpdzukcYWC1xD-vidMZf2uilGkor

* **Quick recap on previous episode**
* **Walkthrough the Loyalty service from the end customer point of view**
* **Async Ingesting Loyalty points for confirmed bookings via Amazon SNS and AWS Lambda**
* **Testing the Ingestion Loyalty function**
    - Discussed patterns to test Typescript functions
    - Walkthrough development setup to test Typescript Lambda functions
    - Walkthrough Ports/Adapters and Dependency injection patterns to ease testing

### Resources

* [Aleksandar Simovic: Lambda Payment Stripe w/ Ports/Adapter patterns](https://github.com/simalexan/api-lambda-stripe-charge)
* [AWS SDK Mock utility](https://www.npmjs.com/package/aws-sdk-mock)
* [Jeremy Daly: Lambda API repo w/ Test patterns](https://github.com/jeremydaly/lambda-api)
* [Gist for Typescript development setup](https://gist.github.com/heitorlessa/a087f4394b38562e1a0aa128386b38b8)
* [macOS Alfred Workflow - AWS Console Services](https://github.com/rkoval/alfred-aws-console-services-workflow)

### Thoughts on decisions we made

**Why Webpack?**

For this project, Webpack gives us the ability to follow [Serverless Best Practices](http://bit.ly/serverless_lens) to easily code single purpose functions while allowing our functions to remain slim enough for deployment.

Webpack also makes it easier to locally prototype our functions in Typescript without having to make any changes in the deployment process, and more importantly it supports organic growth of our functions without having to make any changes in the development tooling setup.

**Why Typescript?**

To this date, this project mainly used Apache Velocity, and Python for the business logic. Typescript brings a new perspective into the development setup that will be useful to customers in the future, and more importantly it enforces the view that it doesn't matter what language we choose for the business logic from the consumer POV + deployment process.

---

### TODO

* [x] Complete Loyalty Unit test
    - [x] Add Unit test for Get
    - [x] Complete Unit test for Ingest (handler, and Dynamo exception)
* [x] Improve Typescript builds
    - Current artifact is including both .ts and .js as well as dependencies
* [x] Document functions
     - [x] Add JSDoc for Get
* [x] Separate business logic from handler
    - [x] Create function
    - [x] Get function
* [x] Update `local-env-vars.json` with latest info from functions
* [x] Enhance Outputs
    - Add API Gateway endpoint (e.g. /{stage}/loyalty)
* [x] Add Stage parameter in SAM template
* [x] Add Loyalty Endpoint URL as Parameter Store
    - /service/loyalty/api-endpoint/${Stage} (../loyalty)
* [x] Add Webpack
* [x] Add source map and test step through debugging with SAM
* [x] Enforce AWS_IAM as Auth for API
* [x] Update GraphQL Schema to include Loyalty Type
* [x] Update GraphQL Schema to include Loyalty Queries
* [x] Front-end: Integrate with Loyalty back-end
* [x] Automate Loyalty AppSync Data source and Resolvers
* [x] Add VSCode Debug configuration for step through debugging
* [x] Front-end: remove membershipNumber
* [x] Add preffix for DynamoDB Table name
* [x] Resolver: Response template should only return Loyalty Points if customerId is the same
    - Exception to "Admin" group
* [x] Update Booking State Machine SNS Payload
* [x] Extend `get` function to include remaining points as part of response
* [x] Update `get` function tests
* [ ] [Optional] Update GraphQL Schema to include Loyalty Mutations
* [ ] [Optional] Update GraphQL Schema to include Loyalty Subscriptions
* [ ] [Optional] Back-end: Implement Local Resolver for Loyalty for reactivity
    - Add mutation to addPoints
    - Add subscriptions onPointsAdded
    - Local Resolver reasoning instead of Dynamo is that Loyalty svc already handles Dynamo integration
* [ ] [Optional] Back-end: Implement Pagination (get/index.ts)
    - Data won't increase at this stage and we can do this outside episode schedule

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
